# Assignment 1
## Exercise 2
### Tasks to Chapter 1 of the Think-Python book

#### 1.9) Ex1)

1. In a print statement, what happens if you leave out one of the parentheses, or both?

In [1]:
print "asd"

SyntaxError: Missing parentheses in call to 'print'. Did you mean print("asd")? (<ipython-input-1-3f2b1fa99d20>, line 1)

Since we are using python3, we cannot natively support non-parenthesis print calls (print is not a function in python2).

2. If you are trying to print a string, what happens if you leave out one of the quotation marks, or both?

In [2]:
print(asd)

NameError: name 'asd' is not defined

It will try to print a variable called 'asd'.

In [3]:
print('asd)

SyntaxError: EOL while scanning string literal (<ipython-input-3-ce83d0ef0ffd>, line 1)

And syntax error if one missing quote.

3. You can use a minus sign to make a negative number like -2. What happens if you put a plus sign before a number? What about 2++2?

In [4]:
2++2

4

'+' is an operator; if binary (elements before and after), it will perform an addition; it can also be unary. In case of -2, it's the unary '-' operator. In case of 2++2, operator precedence will be taken into account, and it will be done in the following order: (2 + (+2)) (parenthesis with logical highest precedence). Therefore, the result is 4.

4. In math notation, leading zeros are ok, as in 02. What happens if you try this in Python?

In [5]:
a = 010


SyntaxError: invalid token (<ipython-input-5-26599968dd8a>, line 1)

In python 2, leading zero means it's an octal number (like 0x is hex), and in python 3 it is not allowed, since "Note that leading zeros in a non-zero decimal number are not allowed. This is for disambiguation with C-style octal literals, which Python used before version 3.0" (from [this](https://docs.python.org/3/reference/lexical_analysis.html) link to the official python docuentation).

5. What happens if you have two values with no operator between them?

In [6]:
123 123

SyntaxError: invalid syntax (<ipython-input-6-86a2e4b7ecb1>, line 1)

Assuming that the question refers to a space between values, we get a syntax error.

#### 1.9) Ex2)

1. How many seconds are there in 42 minutes 42 seconds?

In [7]:
42 * 60 + 42

2562

2562 seconds.

2. How many miles are there in 10 kilometers? Hint: there are 1.61 kilometers in a mile.

In [8]:
10 / 1.61

6.211180124223602

There are ~6.21 miles.

3. If you run a 10 kilometer race in 42 minutes 42 seconds, what is your average pace (time per mile in minutes and seconds)? What is your average speed in miles per hour?

In [9]:
# Rule of 3
# 6.21 miles --- 2562 seconds
# X miles --- 3600 seconds (1 hour)
3600 * 6.21 / 2562

8.725995316159251

Pace is ~8.73 MPH.

### Tasks to Chapter 2 of the Think-Python book

#### 2.10) Ex1)

1. We’ve seen that n = 42 is legal. What about 42 = n?

In [31]:
42 = n

SyntaxError: can't assign to literal (<ipython-input-31-f0c37ed646fc>, line 1)

Illegal to assign a variable name that starts with a number.

2. How about x = y = 1?

In [33]:
x = y = 1
x

1

Legal, since = is an assignment operator, and functions like normal (binary) operators (in this case, from right to left).

3. In some languages every statement ends with a semi-colon, ;. What happens if you put a semi-colon at the end of a Python statement?

In [35]:
print('asd');
print('asd');

asd
asd


It works as expected since python does not *require* semicolons.

4. What if you put a period at the end of a statement?

In [41]:
print('asd').
# 

SyntaxError: invalid syntax (<ipython-input-41-4bb7e12e2f7b>, line 1)

Syntax error, since (depending on the statement), it accesses something from the left element of the period; however, a right element is always needed with the period.

5. In math notation you can multiply x and y like this: x y. What happens if you try that in Python?

In [42]:
a = 2
b = 3
a b

SyntaxError: invalid syntax (<ipython-input-42-dd5f69d9b07f>, line 3)

Syntax error, must use an operator between (in this case, a multiplication binary operator '\*').

#### 2.10) Ex2)

1. The volume of a sphere with radius r is 4/3 π r3. What is the volume of a sphere with radius 5?

In [44]:
r = 5
4/3 * 3.1415 * r**3

523.5833333333333

The volume is ~523.58.

2. Suppose the cover price of a book is \$24.95, but bookstores get a 40% discount. Shipping costs \$3 for the first copy and 75 cents for each additional copy. What is the total wholesale cost for 60 copies?

In [51]:
price = 24.95 * 0.6 # Including discocunt
initialShippingCost = 3
additionalShippingCost = 0.75

price+initialShippingCost + 59*(price+additionalShippingCost)

945.4499999999999

Total wholesale is $945.44 (rounding down which makes sense for a sale, and assuming that the discount is only on the book price, and not in the total amount including the shipping).

3. If I leave my house at 6:52 am and run 1 mile at an easy pace (8:15 per mile), then 3 miles at tempo (7:12 per mile) and 1 mile at easy pace again, what time do I get home for breakfast?

In [66]:
# NOTE: Since this (and every exercise in these chapters) is a small snippet, I will not be using correect configurable variable names nor positions in code. If program was bigger and needs to be configurable, I would adapt to it, but it makes no sense now.
initialTime = 6 * 60 + 52 # Time in minutes from 00:00hs
milesRunEasy = 2
milesRunFast = 3
# 1 mile  --- 8:15
# X miles --- 60:00 (to calculate MPH)
speedEasy = 60 * 60 / (8 * 60 + 15) # in MPH
speedFast = 60 * 60 / (7 * 60 + 12)
print("Easy MPH: " + str(speedEasy))
print("Fast MPH: " + str(speedFast))

finalTime = initialTime + milesRunEasy * speedEasy + milesRunFast * speedFast
finalTimeH = int(finalTime // 60)
finalTimeM = 60 * (finalTime / 60 % finalTimeH) / 100 # Rest of division of hours, base change to time
print("Time to breakfast: " + str(finalTimeH) + ":" + str(int(finalTimeM * 100))) # Manually runding minutes down

Easy MPH: 7.2727272727272725
Fast MPH: 8.333333333333334
Time to breakfast: 7:31


(Please see comments on code to undestand what was done).

#### Tasks to Chapter 3 of the Think-Python book

#### 3.14) Ex1)

Write a function named right_justify that takes a string named s as a parameter and prints the string with enough leading spaces so that the last letter of the string is in column 70 of the display.


right_justify('monty')
                                                                 <div style="text-align: right"> monty </div>
Hint: Use string concatenation and repetition. Also, Python provides a built-in function called len that returns the length of a string, so the value of len('monty') is 5.



In [30]:
1

1

res

#### 3.14) Ex2)

A function object is a value you can assign to a variable or pass as an argument. For example, do_twice is a function that takes a function object as an argument and calls it twice:
<pre>
def do_twice(f):
    f()
    f()
</pre>
Here’s an example that uses do_twice to call a function named print_spam twice.

<pre>
def print_spam():
    print('spam')

do_twice(print_spam)
</pre>

1. Type this example into a script and test it.

In [19]:
1

1

res

2. Modify do_twice so that it takes two arguments, a function object and a value, and calls the function twice, passing the value as an argument.

In [21]:
1 

1

res

3. Copy the definition of print_twice from earlier in this chapter to your script.

In [22]:
1

1

res

4. Use the modified version of do_twice to call print_twice twice, passing 'spam' as an argument.

In [24]:
1 

1

res

5. Define a new function called do_four that takes a function object and a value and calls the function four times, passing the value as a parameter. There should be only two statements in the body of this function, not four.

In [26]:
1

1

res

#### 3.14) Ex3)
Note: This exercise should be done using only the statements and other features we have learned so far.


1. Write a function that draws a grid like the following:

<pre>
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
</pre>

Hint: to print more than one value on a line, you can print a comma-separated sequence of values:

<pre>
print('+', '-')
By default, print advances to the next line, but you can override that behavior and put a space at the end, like this:

print('+', end=' ')
print('-')
The output of these statements is '+ -' on the same line. The output from the next print statement would begin on the next line.
</pre>


In [27]:
1

1

res

2. Write a function that draws a similar grid with four rows and four columns.

In [29]:
1

1

res